File: ai_and_analytics_tools/migrate_prefixes.js
// tools/migrate_prefixes.js
// Run: node tools/migrate_prefixes.js
// Requires: firebase-admin SDK configured with your project credentials
//
// This script adds `searchPrefixes` and `nameLower` to all existing
// product documents. It batches in groups of 500 (Firestore max).

const admin = require('firebase-admin');
admin.initializeApp();
const db = admin.firestore();

const STOP_WORDS = new Set([
    'de', 'of', 'the', 'and', 'for', 'a', 'an', 'eau',
    'by', 'le', 'la', 'les', 'du', 'des', 'en', 'et',
    'or', 'with', 'from',
]);

function buildSearchPrefixes(name) {
    const cleaned = name
        .toLowerCase()
        .replace(/[^\w\s]/g, '')
        .replace(/\s+/g, ' ')
        .trim();

    const words = cleaned.split(' ').filter(w => w.length >= 2 && !STOP_WORDS.has(w));

    const prefixes = new Set();
    for (const word of words) {
        for (let i = 2; i <= word.length; i++) {
            prefixes.add(word.substring(0, i));
        }
    }

    return [...prefixes];
}

async function migrate() {
    const snapshot = await db.collection('products').get();
    const docs = snapshot.docs;
    console.log(`Total products: ${docs.length}`);

    for (let i = 0; i < docs.length; i += 500) {
        const batch = db.batch();
        const chunk = docs.slice(i, i + 500);

        chunk.forEach(doc => {
            const name = doc.data().name || '';
            batch.update(doc.ref, {
                nameLower: name.toLowerCase(),
                searchPrefixes: buildSearchPrefixes(name),
            });
        });

        await batch.commit();
        console.log(`Migrated ${Math.min(i + 500, docs.length)} / ${docs.length}`);
    }

    console.log('✅ Migration complete!');
}

migrate().catch(console.error);

File: mobile_app/lib/features/products/data/migration/migrate_prefixes.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import 'package:perfume_app/features/products/data/models/product_model.dart';

/// One-time migration: adds `searchPrefixes` and `nameLower` to all
/// existing product documents in Firestore.
///
/// Call this ONCE, then remove or disable it.
/// Usage: await migrateProductPrefixes();
Future<void> migrateProductPrefixes() async {
  final firestore = FirebaseFirestore.instance;
  final snapshot = await firestore.collection('products').get();

  debugPrint('🔄 Migrating ${snapshot.docs.length} products...');

  // Firestore batch limit is 500
  final batches = <WriteBatch>[];
  WriteBatch currentBatch = firestore.batch();
  int count = 0;

  for (final doc in snapshot.docs) {
    final name = doc.data()['name'] ?? '';
    final prefixes = buildSearchPrefixes(name);

    currentBatch.update(doc.reference, {
      'nameLower': name.toString().toLowerCase(),
      'searchPrefixes': prefixes,
    });

    count++;
    if (count % 500 == 0) {
      batches.add(currentBatch);
      currentBatch = firestore.batch();
    }
  }

  // Add the last batch if it has remaining items
  if (count % 500 != 0) {
    batches.add(currentBatch);
  }

  // Commit all batches
  for (int i = 0; i < batches.length; i++) {
    await batches[i].commit();
    debugPrint('✅ Batch ${i + 1}/${batches.length} committed');
  }

  debugPrint('🎉 Migration complete! $count products updated.');
}
